scipy/special/xsf/cephes/trig.h

```diff
@@ -19,40 +20,66 @@ namespace cephes {
     /* Compute sin(pi * x). */
     template <typename T>
     XSF_HOST_DEVICE T sinpi(T x) {
-        T s = 1.0;
-
-        if (x < 0.0) {
-            x = -x;
-            s = -1.0;
-        }
-
-        T r = std::fmod(x, 2.0);
-        if (r < 0.5) {
-            return s * std::sin(M_PI * r);
-        } else if (r > 1.5) {
-            return s * std::sin(M_PI * (r - 2.0));
-        } else {
-            return -s * std::sin(M_PI * (r - 1.0));
-        }
+        T s = std::copysign(1.0, x);
+        x = std::abs(x);
+
+        T r = x - std::ldexp(std::floor(std::ldexp(x, -1)), 1);
+        s = r <= 1.0 ? s : -s;
+        r = r - std::floor(r);
+        r = r <= 0.5 ? r : (1.0 - r);
+
+        T y = std::copysign(std::sin(M_PI * r), s);
+ 
+        return y;
     }
 
     /* Compute cos(pi * x) */
     template <typename T>
     XSF_HOST_DEVICE T cospi(T x) {
-        if (x < 0.0) {
-            x = -x;
-        }
-
-        T r = std::fmod(x, 2.0);
-        if (r == 0.5) {
-            // We don't want to return -0.0
-            return 0.0;
-        }
-        if (r < 1.0) {
-            return -std::sin(M_PI * (r - 0.5));
-        } else {
-            return std::sin(M_PI * (r - 1.5));
-        }
+        x = std::abs(x);
+
+        T r = x - std::ldexp(std::floor(std::ldexp(x, -1)), 1);
+        r = r <= 1.0 ? (0.5 - r) : (r - 1.5);
+
+        T y = std::sin(M_PI * r);
+
+        return y;
+    }
+
+    /* Compute tan(pi * x) */
+    template <typename T>
+    XSF_HOST_DEVICE T tanpi(T x) {
+        T s = std::copysign(1.0, x);
+        x = std::abs(x);
+
+        T r = x - std::ldexp(std::floor(std::ldexp(x, -1)), 1);
+        s = r <= 1.0 ? s : -s;
+
+        T rs = r - std::floor(r);
+        rs = rs <= 0.5 ? rs : (1.0 - rs);
+        T rc = r <= 1.0 ? (0.5 - r) : (r - 1.5);
+
+        T y = std::copysign(std::sin(M_PI * rs), s) / std::sin(M_PI * rc);
+
+        return y;
+    }
+
+    /* Compute cot(pi * x) */
+    template <typename T>
+    XSF_HOST_DEVICE T cotpi(T x) {
+        T s = std::copysign(1.0, x);
+        x = std::abs(x);
+
+        T r = x - std::ldexp(std::floor(std::ldexp(x, -1)), 1);
+        s = r <= 1.0 ? s : -s;
+
+        T rs = r - std::floor(r);
+        rs = rs <= 0.5 ? rs : (1.0 - rs);
+        T rc = r <= 1.0 ? (0.5 - r) : (r - 1.5);
+
+        T y = std::sin(M_PI * rc) / std::copysign(std::sin(M_PI * rs), s);
+
+        return y;
```

Comment:
I wonder if it might make sense to add sniff tests for the new functions, `cotpi` and `tanpi`, and perhaps to augment the tests for the original two functions that were apparently not standard compliant?

Reply:
@tylerjereddy 
Thank you for your comment.

In other words, would the following code be acceptable in the hope that the compiler will optimize it?

```cpp
    /* Compute tan(pi * x) */
    template <typename T>
    XSF_HOST_DEVICE T tanpi(T x) {
        T y = sinpi(x) / cospi(x);

        return y;
    }

    /* Compute cot(pi * x) */
    template <typename T>
    XSF_HOST_DEVICE T cotpi(T x) {
        T y = cospi(x) / sinpi(x);

        return y;
    }
```

Reply:
Maybe lines are getting crossed, I wasn't suggesting you change your source, I was just talking about potentially writing some regression tests in a separate file. Maybe the `special` folks will have more to say about that.